<h1> Quiz 1 : Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa kalian masing2?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matriks? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

Jawab :
1. Data latih adalah data yang digunakan pada Machine Learning untuk mendapatkan nilai weight, sehingga Machine Learning dapat memprediksikan sesuatu. Data validasi biasanya bagian dari dataset yang dipisahkan dari data training untuk melakukan pengukuran pada model Machine Learning sehingga dapat mengetahui loss dan akurasi dari proses training. Data test adalah data baru yang digunakan untuk menggunakan model Machine Learning yang telah selesai proses training
2. Dapat diperhatikan melalui akurasi validasi dan loss validasi hasil training model.
3. Salah satu metrics untuk mengetahui pengecekan setiap pada untuk setiap kelas yang telah di training
4. Fungsi untuk memunculkan nilai Precision, recall, dan akurasi untuk setiap kelas.

<h1>Quiz 2 : Applikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payu dara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut : https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien. dengan ketentuan sebagai berikut :

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [ ]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [ ]:
x = df.drop('survival_status', axis=1)
y = df['survival_status']

In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.neighbors import KNeighborsClassifier

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=14)
model = KNeighborsClassifier(n_neighbors = 35, weights ='distance')
model.fit(x_train,y_train)
h_cv = cross_val_score(model, x_test, y_test, cv = 14, scoring='roc_auc')
print(h_cv.mean())
y_pred = model.predict(x_test)
print(confusion_matrix(y_test,y_pred,labels = [1,2]))
print(classification_report(y_test,y_pred))

0.6458333333333333
[[48  6]
 [14  9]]
              precision    recall  f1-score   support

           1       0.77      0.89      0.83        54
           2       0.60      0.39      0.47        23

    accuracy                           0.74        77
   macro avg       0.69      0.64      0.65        77
weighted avg       0.72      0.74      0.72        77



7. Sangat baik, dapat mencapai akurasi 83%

8. Kurang baik, karena hanya dapat mencapai akurasi 47%